iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/infospace/ImportInformationSpaceJobUtils.java

```diff
@@ -0,0 +1,70 @@
+     * @param sourceDataset              {@link eu.dnetlib.dhp.schema.oaf.Dataset} RDD.
+     * @param sourceOtherResearchProduct {@link eu.dnetlib.dhp.schema.oaf.OtherResearchProduct} RDD.
+     * @param sourcePublication          {@link eu.dnetlib.dhp.schema.oaf.Publication} RDD.
+     * @param sourceSoftware             {@link eu.dnetlib.dhp.schema.oaf.Software} RDD.
+     * @param spark                      Instance of SparkSession.
+     * @return An RDD of {@link IdentifierMapping} with original ids of {@link eu.dnetlib.dhp.schema.oaf.Result} records
+     * replaced.
+     */
+    public static JavaRDD<IdentifierMapping> mapToObjectStoreId(JavaRDD<IdentifierMapping> dedupRelation,
+                                                                JavaRDD<eu.dnetlib.dhp.schema.oaf.Dataset> sourceDataset,
+                                                                JavaRDD<eu.dnetlib.dhp.schema.oaf.OtherResearchProduct> sourceOtherResearchProduct,
+                                                                JavaRDD<eu.dnetlib.dhp.schema.oaf.Publication> sourcePublication,
+                                                                JavaRDD<eu.dnetlib.dhp.schema.oaf.Software> sourceSoftware,
+                                                                SparkSession spark) {
+        Dataset<eu.dnetlib.dhp.schema.oaf.Dataset> sourceDatasetDS = spark.createDataset(
+                sourceDataset.rdd(), Encoders.bean(eu.dnetlib.dhp.schema.oaf.Dataset.class));
+        Dataset<eu.dnetlib.dhp.schema.oaf.OtherResearchProduct> sourceOtherResearchProductDS = spark.createDataset(
+                sourceOtherResearchProduct.rdd(), Encoders.bean(eu.dnetlib.dhp.schema.oaf.OtherResearchProduct.class));
+        Dataset<eu.dnetlib.dhp.schema.oaf.Publication> sourcePublicationDS = spark.createDataset(
+                sourcePublication.rdd(), Encoders.bean(eu.dnetlib.dhp.schema.oaf.Publication.class));
+        Dataset<eu.dnetlib.dhp.schema.oaf.Software> sourceSoftwareDS = spark.createDataset(
+                sourceSoftware.rdd(), Encoders.bean(eu.dnetlib.dhp.schema.oaf.Software.class));
+
+        org.apache.spark.sql.Dataset<Row> resultIdMapDF = sourceDatasetDS.select(col("id"), explode(col("originalId")).as("oid"))
+                .union(sourceOtherResearchProductDS.select(col("id"), explode(col("originalId")).as("oid")))
+                .union(sourcePublicationDS.select(col("id"), explode(col("originalId")).as("oid")))
+                .union(sourceSoftwareDS.select(col("id"), explode(col("originalId")).as("oid")))
+                .distinct();
+
+        org.apache.spark.sql.Dataset<Row> dedupRelationDF = new AvroDatasetSupport(spark).toDF(
+                spark.createDataset(dedupRelation.rdd(), Encoders.kryo(IdentifierMapping.class)), IdentifierMapping.SCHEMA$);
+
+        Dataset<Row> dedupRelationWithOidDF = dedupRelationDF
```

Comment:
Just as we have already discussed today: we should change the joining order because we need to translate all the identifiers in `DocumentContentUrl` down the stream, not only deduped ones. So we need to have all entries from originalId->persistentId from the graph supplemented with persistend->dedupedId mapping wherever applicable (for each persistentId from the originalId->perisitentId map which could be found in persistend->dedupedId map).

More in:
https://issue.openaire.research-infrastructures.eu/issues/6510#note-24

Reply:
Yes, you're right I didn't get that at the beginning. But giving this more thought I think that just a change in join order is not enough because we can have three types of id mappings as the result:
* for deduplicated entities that are not `Result` types we should map using deduplication mapping only - this case is the same as what we currently do
* for `Result` types that are not deduplicated we should map using `Result` mapping only - in this case we should map using `id` and `originalId` fields of the `Result` entity
* for deduplicated entities of `Result` types  we should map using a combination of deduplication mapping and `Result` mapping - in this case we should map using `newId` field of deduplication mapping and `originalId` field of `Result` mapping

This means that we end up with a full outer join between both mappings with additional selection of proper 'ouput' fields based on the join result. Running this version of the code with recent PROD run results gives 95.347.794 total mapping records with 61.114.481 records matching the first case (only deduplication mapping), 18.736.545 records matching the second case (only `Result` mapping) and 15.496.768 records matching the last case (both mappings). I've updated the code to implement this change.

Secondly, I've added filtering to the `Result` mapping based on entity `id` field, because (strangely) I forgot to do that. I've also changed the name of the method because the old name was misleading.